feat: log scale and step size control for range filters [WIP]

stevetracvc opened the pull request:
### SUMMARY
This is my PR for issue Extra features for Range Filters #16582 (sorry, I screwed up my original PR #16591)

Range filters are fantastic, but there is no simple way to adjust the step size or the scale of the range. Scale is useful for large ranges, eg a log scale, so that you can easily select small and large values

- Step Size is just an additional Advanced Config dropdown box with several pre-populated choices, as well as a free form option. The tricky part was getting it to properly work in the Native Filter Config modal, since they're expecting all parameters to be checkboxes.
- Scale is a little trickier, but it could possibly be set up as multiple options (eg log, power, etc). Currently I implemented a log scale option, which would make zeros hard. I'm currently ignoring that :)


### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Config Screen Usage
![config screen use](https://user-images.githubusercontent.com/70416691/132033938-6d095135-698b-448c-aeb7-2ec359388038.gif)

Log Scale with 0.01 step size
![log-scale and 0 01 step](https://user-images.githubusercontent.com/70416691/132033942-06eda7eb-1f09-49c9-a3bf-847de4848a5e.gif)

Linear Scale with 100k step size
![linear-scale and 100k step](https://user-images.githubusercontent.com/70416691/132033945-2b592134-3614-4ac4-84eb-d6eb497e9e36.gif)

Config Screen Defaults
![config screen defaults](https://user-images.githubusercontent.com/70416691/132033950-9f026b6b-b364-465a-81c1-c224869b4c65.png)



### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: Fixes #16582
- [x] Required feature flags: DASHBOARD_NATIVE_FILTERS, DASHBOARD_CROSS_FILTERS
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### Known Issues
- currently, I add 1 to the min and max values if it's supposed to be log scale. I probably need to check if (a) 0 is one of the min/max values, and (b) whether the range is negative

stevetracvc committed:
feat: log scale and step size control for range filters

stevetracvc committed:
fix: prettier

stevetracvc committed:
fix: dupe min/max props

stevetracvc committed:
fix: boolean math issue

stevetracvc committed:
fix: doubled up in controlPanel

geido commented on superset-frontend/src/filters/components/Range/controlPanel.ts:
```diff
@@ -69,6 +69,41 @@ const config: ControlPanelConfig = {
+            name: 'logScale',
+            config: {
+              type: 'CheckboxControl',
+              label: t('Logarithmic Scale'),
```
```suggestion
              label: t('Logarithmic scale'),
```

geido commented on superset-frontend/src/filters/components/Range/controlPanel.ts:
```diff
@@ -69,6 +69,41 @@ const config: ControlPanelConfig = {
+            name: 'stepSize',
+            config: {
+              type: 'SelectControl',
+              label: t('Step Size'),
```
```suggestion
              label: t('Step size'),
```

geido commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/getControlItemsMap.tsx:
```diff
@@ -211,6 +213,69 @@ export default function getControlItemsMap({
+                clearable={false}
+                freeForm={controlItem.config.freeForm}
+                disabled={controlItem.config.affectsDataMask && disabled}
+                onChange={(value: any) => {
```
Do you think we can be more specific for the type of the `value` of this `onChange` event?

stevetracvc replied:
Good question @geido
So the Native Filter control didn't previously allow for a drop down, so as a new feature it's hard to imagine hoe else this could be used. I'm assuming string and number are likely the only types here, and I just forgot to change it from any.

stevetracvc replied:
I can see changing the "logarithmic scale" from a checkbox to a drop down, with choices for linear, power, log, symlog (would need an input box for the symlog constant), maybe others. I'm using d3-scale, which has those built in:
https://github.com/d3/d3-scale#continuous-scales

geido commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/getControlItemsMap.tsx:
```diff
@@ -211,6 +213,69 @@ export default function getControlItemsMap({
+                    [controlItem.name]: value,
+                    defaultDataMask: null,
+                  });
+                  // controlItem.config.value = value;
```
```suggestion
```

geido reviewed (commented):
Hey @stevetracvc I appreciate a lot the contributions that you are making! I am reviewing this PR which is still a WIP for some quick nits. Let me know when you are ready for a more in-depth review. Thank you!

rusackas commented:
Fixes #17043

geido commented:
@stevetracvc I see that some tests are requiring some love. Let me know how I can help to move this forward! Thank you

stevetracvc committed:
merged in master 1.4.1ish

stevetracvc committed:
fixes to default/reset value and options for scaling

stevetracvc committed:
fix: prettier stuff

stevetracvc committed:
fix: more prettier

stevetracvc committed:
fix: once more

stevetracvc committed:
fix: test needed scaling value

stevetracvc committed:
Fixed issue with reset filters and added cube/quad roots

--ran into an issue when pulling and pushing some of this older code,
now it should work when resetting the filter

--it's easy to add in extra transforms, so I added cube and quad root

stevetracvc committed:
added tests for all scaling functions

stevetracvc committed:
fix: prettier

stevetracvc committed:
fix: foreach issue

stevetracvc committed:
fix: cleanup old code

stevetracvc committed:
fix: issue with default scaling function

fixed default scaling in controlPanel, and added default in
RangeFilterPlugin to handle importing old range filters to the
new scaling system

geido commented:
/testenv up

kgabryje commented:
Thanks for the improvement @stevetracvc!
1 issue that I spotted while testing - if the upper range value is smaller then `current value + step`, then we can't select it. For example, in the video my range is about [0, 41]. When I select step 25, then I can only select values 0 and 25. I think 41 should also be selectable, even if that would mean making smaller step. WDYT?

https://user-images.githubusercontent.com/15073128/161023937-9bfd852a-5ffd-4b0c-8c66-dd30da52d725.mov

kgabryje commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/getControlItemsMap.tsx:
```diff
@@ -214,6 +216,69 @@ export default function getControlItemsMap({
+    .filter(
+      (controlItem: CustomControlItem) =>
+        controlItem?.config?.renderTrigger &&
+        controlItem.name !== 'sortAscending' &&
```
We probably don't need this check?

stevetracvc replied:
You mean and just leave it to checking if it's a SelectControl? I'm fine with that but honestly I never checked what those other conditions are for. I just duplicated the previous section and tweaked it for a select :laughing:

stevetracvc replied:
renderTrigger is needed, otherwise the COLUMN select box shows up again. sortAscending isn't though.

kgabryje commented:
We might want to rearrange the controls. Clicking on "Filter has default value" below the step controls triggers displaying default value control above the step controls, which might be confusing. Also, the values chosen in step controls affect how the default value control works, which would probably mean that step controls should be placed above default value control
<img width="631" alt="image" src="https://user-images.githubusercontent.com/15073128/161026010-8ac3b3d5-f271-44bb-9581-4bfc5233655e.png">

CC @kasiazjc

stevetracvc commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/getControlItemsMap.tsx:
```diff
@@ -149,6 +150,7 @@ export default function getControlItemsMap({
       (controlItem: CustomControlItem) =>
         controlItem?.config?.renderTrigger &&
         controlItem.name !== 'sortAscending' &&
+        controlItem?.config?.type !== 'SelectControl' &&
```
Why a checkbox? I added a whole new section (below) to handle dropdown selects because that doesn't exist anywhere else in the config modal. It's for the step size and scaling function select boxes

stevetracvc replied:
Oh I reread this and get it now. I guess the question is, how flexible should these controls be? And since everything else is a checkbox, no one bothered to check before. I'll make that change, thanks!

stevetracvc commented:
> We might want to rearrange the controls. Clicking on "Filter has default value" triggers displaying default value control above the step controls, while step controls affect how default value picker works - that probably means that step controls should be above default value. <img alt="image" width="631" src="https://user-images.githubusercontent.com/15073128/161026010-8ac3b3d5-f271-44bb-9581-4bfc5233655e.png">
> 
> CC @kasiazjc

That's an interesting observation but I have no idea how to do that. The default value stuff code is handled somewhere else, and honestly I'm not sure how it decides on the location of controlPanel components. In the controlPanel code for the Range Filter, I have

```
    {
      label: t('UI Configuration'),
      expanded: true,
      controlSetRows: [
        [
          {
            name: 'enableEmptyFilter',
            config: {
              type: 'CheckboxControl',
              label: t('Filter value is required'),
              default: false,
              renderTrigger: true,
              description: t(
                'User must select a value before applying the filter',
              ),
            },
          },
          {
            name: 'enableSingleValue',
            config: {
              type: 'CheckboxControl',
              label: t('Single value'),
              default: SingleValueType.Exact,
              renderTrigger: true,
              description: t('Use only a single value.'),
            },
          },
        ],
        [
          {
            name: 'scaling',
            config: {
              type: 'SelectControl',
              label: t('Scaling Function'),
              default: PluginFilterRangeScalingFunctions.LINEAR,
              renderTrigger: true,
              freeForm: false,
              choices: Object.keys(
                SCALING_FUNCTION_ENUM_TO_SCALING_FUNCTION,
              ).map(key => [
                key,
                SCALING_FUNCTION_ENUM_TO_SCALING_FUNCTION[key].display,
              ]),
              description: t('Choose a scaling function for the slider.'),
            },
          },
          {
            name: 'stepSize',
            config: {
              type: 'SelectControl',
              label: t('Step Size'),
              default: 1,
              renderTrigger: true,
              freeForm: true,
              choices: [
                [0.001, 0.001],
                [0.01, 0.01],
                [0.1, 0.1],
                [1, 1],
                [2, 2],
                [10, 10],
                [25, 25],
                [100, 100],
              ],
              description: t('Set the slider step size.'),
            },
          },
        ],
      ],
    },
```
But you can clearly see that's not the order the modal decides to put stuff.

stevetracvc commented:
> Thanks for the improvement @stevetracvc! 1 issue that I spotted while testing - if the upper range value is smaller then `current value + step`, then we can't select it. For example, in the video my range is about [0, 41]. When I select step 25, then I can only select values 0 and 25. I think 41 should also be selectable, even if that would mean making smaller step. WDYT?
> 
>  Screen.Recording.2022-03-31.at.11.29.09.mov

I do think there are some issues with endpoints every now and then. I thought I fixed it for things like that (honestly I've never used anything bigger than 1 :laughing: ) especially with, eg, log scaling, when trying to drag to the end. Those were a rounding issue but this one isn't. I'll look around and see what I can figure out. The tests warn about the endpoints not being increments of the step size. What do you think would be the correct approach? Should the endpoint be **moved**? In your example, it would get moved to 50.

kgabryje commented:
> Should the endpoint be **moved** to an increment of the step size? In your example, it would get moved to 50.
I'd probably expect something like `min(maxRangeValue, currentValue + step)`, so for example with linear scale, step 25 and values in range [0,41], possible selections would be 0, 25, 41. But to be honest, the current behaviour _kinda_ makes sense too... @yousoph @jess-dillard what do you think?

kgabryje commented:
> But you can clearly see that's not the order the modal decides to put stuff. enableEmptyFilter is the last control in the modal, yet the first in the controlSetRows array. I still don't get exactly how that controlSetRows should be structured, as I would have thought the enableEmptyFilter and enableSingleValue controls would be on the same row (since they're in the same controlSetRow) which is (I believe) how it works for visualization controlPanel configs.

I'm not very familiar with native filters control panels... @michael-s-molina @geido do you guys have some context on this?

jess-dillard commented:
> 

@kgabryje I agree with you – I would also expect the possible selections to be 0, 25, and 41 in that scenario.

stevetracvc commented on superset-frontend/src/filters/components/Range/RangeFilterPlugin.tsx:
```diff
@@ -186,41 +218,62 @@ export default function RangeFilterPlugin(props: PluginFilterRangeProps) {
+          inverseScale(upper),
+        ),
         filterState: {
           value: lower !== null || upper !== null ? value : null,
```
What does the filterState.value do? The variable "value" is transformed, meaning it could be the log of the actual desired value. So I'd expect that this would need to actually be
`value.map(inverseScale)`
like how it is below for the lower and upper vars. But the queries appear to run properly, and the text labels are correct.... :thinking:

geido commented:
> > We might want to rearrange the controls. Clicking on "Filter has default value" triggers displaying default value control above the step controls, while step controls affect how default value picker works - that probably means that step controls should be above default value. <img alt="image" width="631" src="https://user-images.githubusercontent.com/15073128/161026010-8ac3b3d5-f271-44bb-9581-4bfc5233655e.png">
> > CC @kasiazjc
> 
> That's an interesting observation but I have no idea how to do that. The default value stuff code is handled somewhere else, and honestly I'm not sure how it decides on the location of controlPanel components. In the controlPanel code for the Range Filter, I have
> 
> ```
>     {
>       label: t('UI Configuration'),
>       expanded: true,
>       controlSetRows: [
>         [
>           {
>             name: 'enableEmptyFilter',
>             config: {
>               type: 'CheckboxControl',
>               label: t('Filter value is required'),
>               default: false,
>               renderTrigger: true,
>               description: t(
>                 'User must select a value before applying the filter',
>               ),
>             },
>           },
>           {
>             name: 'enableSingleValue',
>             config: {
>               type: 'CheckboxControl',
>               label: t('Single value'),
>               default: SingleValueType.Exact,
>               renderTrigger: true,
>               description: t('Use only a single value.'),
>             },
>           },
>         ],
>         [
>           {
>             name: 'scaling',
>             config: {
>               type: 'SelectControl',
>               label: t('Scaling Function'),
>               default: PluginFilterRangeScalingFunctions.LINEAR,
>               renderTrigger: true,
>               freeForm: false,
>               choices: Object.keys(
>                 SCALING_FUNCTION_ENUM_TO_SCALING_FUNCTION,
>               ).map(key => [
>                 key,
>                 SCALING_FUNCTION_ENUM_TO_SCALING_FUNCTION[key].display,
>               ]),
>               description: t('Choose a scaling function for the slider.'),
>             },
>           },
>           {
>             name: 'stepSize',
>             config: {
>               type: 'SelectControl',
>               label: t('Step Size'),
>               default: 1,
>               renderTrigger: true,
>               freeForm: true,
>               choices: [
>                 [0.001, 0.001],
>                 [0.01, 0.01],
>                 [0.1, 0.1],
>                 [1, 1],
>                 [2, 2],
>                 [10, 10],
>                 [25, 25],
>                 [100, 100],
>               ],
>               description: t('Set the slider step size.'),
>             },
>           },
>         ],
>       ],
>     },
> ```
> 
> But you can clearly see that's not the order the modal decides to put stuff. enableEmptyFilter is the last control in the modal, yet the first in the controlSetRows array. I still don't get exactly how that controlSetRows should be structured, as I would have thought the enableEmptyFilter and enableSingleValue controls would be on the same row (since they're in the same controlSetRow) which is (I believe) how it works for visualization controlPanel configs.

@stevetracvc @kgabryje 

We had a conversation with @michael-s-molina who has suggested the following:

- The easiest solution is just to leave it in the current position if we can live with it (of course)
- Another way is to move the code directly into `FiltersConfigForm` between the description and the default value

You are right this wasn't straightforward forward but we aim to improve the codebase of native filters in the upcoming phase 2.

Some other observations that came out from our conversation:

- This functionality should be under a checkbox as it’s very specific, such as “Customise default step scaling function”
- The inputs should not take the full width of the Modal. Please follow the same UI of existing inputs
- If the step size is lower than two decimals, the exact number of decimals should also be visible in the left sidebar, otherwise it’s impossible to see those decimals changing.

Let me know if you have any questions. Thanks for the hard work!

stevetracvc committed:
fix: cleaned up FiltersConfigModal a bit

stevetracvc commented:
Awesome! Questions inline:

> 
> We had a conversation with @michael-s-molina who has suggested the following:
> 
> * The easiest solution is just to leave it in the current position if we can live with it (of course)
> * Another way is to move the code directly into `FiltersConfigForm` between the description and the default value

If we did that, then that moves more of the config out of the specific native filter component's controlPanel file and into the main one. I had tried to make this more general so that other filter types could also include select boxes for controls if needed. If you do want the FiltersConfigForm modified instead, it seems like it should go under the "single value" checkbox, as part of the configuration rather than the settings. What do you think?

> 
> You are right this wasn't straightforward forward but we aim to improve the codebase of native filters in the upcoming phase 2.

Since that's the plan, which route would be better? It seems like we'd want everything in the specific component's controlPanel rather than spread across the two files. And maybe revisit this when phase 2 is complete?

> 
> Some other observations that came out from our conversation:
> 
> * This functionality should be under a checkbox as it’s very specific, such as “Customise default step scaling function”

Is it really that specific? Am I one of the few people using range filters for a large range of values? If you want it hidden like that, then the wording on that needs improvement, since step size and scaling can each be used independent of the other one. Eg, you can't currently use a range filter for floats since the step size is inherently 1 (especially on a range 0 to 1). I feel like each one is its own thing, and step size could be incorporated without bothering with the scaling.


> * The inputs should not take the full width of the Modal. Please follow the same UI of existing inputs

I finally found where it does its magic, so I made them half width now (I haven't pushed yet). I'm starting to understand how everything works here, but a lot of what I originally did was just get it to work because these were features I needed...so I copied and pasted code.

The issue is that FiltersControlModal calls getControlItemsMap, which calls getControlItems, which **flattens** the controlPanel config array. I think ideally, step size and scaling function would be in the same row, don't you think? But that would take a decent rework of the FiltersConfigModal code to make that work.


> * If the step size is lower than two decimals, the exact number of decimals should also be visible in the left sidebar, otherwise it’s impossible to see those decimals changing.

Not sure I follow on this one, though it also might be related to the issue of values not being in exact increments of the step size. When I have the step size at 0.001, linear scale, for something with values 0 to 1, each increment is actually 0.005, which I think is a limitation on the antd slider control.

> 
> Let me know if you have any questions. Thanks for the hard work!

stevetracvc committed:
fix: clarified step size is increments of transformed scale

stevetracvc committed:
fix: issue with using range filter as a dashboard filter

stevetracvc committed:
feat: added validators to FiltersConfigForm loaded from filter's controlPanel

stevetracvc committed:
fix: display precision now linked to stepSize

there was an issue with the displayed values not having enough
significant digits to match the stepSize (SMART_NUMBER formatting)

stevetracvc commented:
Actually I think I get what @geido you meant about the decimals being displayed. SMART_NUMBER formatting was used, which didn't seem to give enough digits of precision. I just changed that to switch to ".nf" where n is the precision of the stepSize.

Also I added validators to the stepSize in the controlPanel, which also required some conversion between the controlPanel validators and the StyledFormItem rules. It seems kind of odd that there are two drastically different validation methods being used. Will visualizations move to the FormItem rules? This is more obvious in the RangeFilter case because it can be both a native filter component AND a dashboard visualization.

stevetracvc committed:
fix: remove direct dependence on rc-field-form

stevetracvc committed:
fix: max value now pushed to an integer increment of stepSize

There were issues with odd combinations of stepSize, min, and max
values. This now pushes the maximum value for the slider up to the
next integer increment of stepSize. Unfortunately, the antd slider
won't allow steps such as (0, 25, 50, 60) so it has to get bumped
to (0, 25, 50, 75)

stevetracvc commented:
> > I'd probably expect something like min(maxRangeValue, currentValue + step), so for example with linear scale, step 25 and values in range [0,41], possible selections would be 0, 25, 41. But to be honest, the current behaviour kinda makes sense too... @yousoph @jess-dillard what do you think?
> 
> @kgabryje I agree with you – I would also expect the possible selections to be 0, 25, and 41 in that scenario.

Unfortunately it looks like this is a restriction on the antd slider. You can't have 0, 25, 41. They say (max - min) / stepSize must be a whole number (evenly divisible). Recent push tackles this by bumping the max up to the next possible increment of stepSize. Not ideal, I know.

stevetracvc committed:
fix: getControlItemsMap conditional on CheckboxControl

stevetracvc committed:
fix: value type for SelectControl onChange

stevetracvc committed:
fix: prev commit requires config type to explicitly be CheckboxControl

stevetracvc committed:
added a couple tests for the new SelectControl in getControlItemsMap

kgabryje commented:
> Unfortunately it looks like this is a restriction on the antd slider. You can't have 0, 25, 41. They say (max - min) / stepSize must be a whole number (evenly divisible). Recent push tackles this by bumping the max up to the next possible increment of stepSize. Not ideal, I know, but 0, 25, 50 is better than just 0, 25, and an unselectable 41

@stevetracvc Ah I see. @yousoph @lauderbaugh if you guys are fine with this then I am too 🙂

kgabryje commented:
/testenv up

stevetracvc commented:
@kgabryje weird. It works for step sizes LESS than 1, but not greater than. Eg on the video game sales, global sales is 0 to 82. Step size of 0.001 on log scale takes it up to 82.560. But if step size is 1, it stops at 81. I'll investigate

rusackas commented:
@stevetracvc it looks like this was on the road to success, but things halted. Do you still have interest/bandwidth to get this through?

stevetracvc commented:
@rusackas I got stuck on something. It works well enough for my usage but not for release. I would probably have time in Feb to look at it again

rusackas commented:
Heya @stevetracvc I'll go ahead and close this since it's been a couple years since it saw action. Please re-open it (or let me know do to so) and we'll be excited to rekindle this when the day comes!
